### DISCUSSION TITLE
Multiple quotation mark replacement fails

### DISCUSSION THREAD

**Discussion by aral** [helix-editor/helix#10369] [General]
### Summary

Replacing multiple quotation marks results in inconsistent behaviour.

## To reproduce

1. Add the following code to a JavaScript page:

```js
export default _ => html`
  <script defer src='https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js'></script>
  <h1 x-data="{ message: 'I ❤️ Alpine'}" x-text='message'></h1>
  <div x-data="{ count: 0 }">
    <button x-on:click="count++" style=>Increment</button>
    <span x-text="count"></span>
  </div>
`
```

2. Select all and then select all double quote marks. `% s"`

    ![image](https://user-images.githubusercontent.com/41798/207350135-7f765505-68df-4d3c-a26b-1e70eeb1e3f0.png)

3. Press <kbd>c</kbd> and type a single quote(`'`)

## What should happen

All double quotes should be replaced with one single quote.

## What actually happens

Some double quotes are replaced with one single quote, others are replaced with two.

![image](https://user-images.githubusercontent.com/41798/207351198-23d38f3e-8a08-480b-b641-3fdff30e40a0.png)

This is also inconsistent with <kbd>r</kbd> which does the right thing.

### Helix log

n/a

### Platform

Linux (Fedora Silverblue 37)

### Terminal Emulator

Black Box

### Helix Version

helix 22.12 (0f2ae35a)

**Comment by CptPotato** [comment]
This is caused by auto pairs of quotes, braces, etc. and I think it's the intended behavior.

If you want to replace `"` quotes to `'` use `r` instead: `%s"<ret>r'`

If the replacement text is longer (not a single character), yank the desired replacement text, then `%s"<ret>R` (note the capital `R` which replaces all selections with the yanked text).

Edit: Regarding the discrepancy between `c` and `r`:
`r` performs a 1:1 character replacement, the number of characters doesn't change. On the other hand, `c` is a way of writing new text where you generally want auto pairs to work.

**Comment by aral** [comment]
@CptPotato I perhaps need to open a bug on auto-pairs then because they’re really not working as they should.

e.g., If I type: <kbd>h</kbd><kbd>t</kbd><kbd>m</kbd><kbd>l</kbd><kbd>`</kbd> in JavaScript to create a tagged template, I get:

```
html`
```

Then, since auto pairs failed, if I type another <kbd>`</kbd>, I get:

```
html``|`
```

(I’ve added the | to show where the cursor is placed. That’s not in the actual output.)

Should I open a separate issue for this?

CC @the-mikedavis

**Comment by the-mikedavis** [comment]
I think that's an instance of https://github.com/helix-editor/helix/issues/4100
